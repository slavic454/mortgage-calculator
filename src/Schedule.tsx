import React, { memo } from 'react';
import { useSelector } from 'react-redux';
import { RootState } from './index';
import Ad from './assets/mortgage-ad.jpg';
import Summary from './Summary';
import { EQUAL_PAYMENTS_STRATEGY } from './formulas/EqualPayments';
import { chooseStrategy } from './formulas';
import ScheduleTable from './ScheduleTable';
import { Card, CardImg } from 'reactstrap';

const INTEREST_RATE = 0.035;

export const MONTHS_IN_YEAR = 12;

export const addCurrency = (money: string) => money + '$';

const Schedule = () => {
  const { amount, payback } = useSelector(({ amount, payback }: RootState) => ({
    amount,
    payback,
  }));
  const paybackTimeInMonths = payback * MONTHS_IN_YEAR;

  const monthlyInterestRate = 1 + INTEREST_RATE / MONTHS_IN_YEAR;

  if (!amount || !payback) {
    return (
      <Card>
        <CardImg src={Ad} />
      </Card>
    );
  }

  const { schedule, totalInterest } = chooseStrategy(EQUAL_PAYMENTS_STRATEGY)({
    amount,
    paybackTimeInMonths,
    interestRate: INTEREST_RATE,
    monthlyInterestRate,
  });

  return (
    <>
      <Summary
        amount={amount}
        totalInterests={totalInterest}
        rate={INTEREST_RATE}
      />
      <ScheduleTable schedule={schedule} />
    </>
  );
};

export default memo(Schedule);
