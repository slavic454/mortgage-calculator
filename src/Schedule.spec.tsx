import React from 'react';
import renderer from 'react-test-renderer';
import Schedule from './Root';

jest.mock('react-redux', () => {
  const ActualReactRedux = jest.requireActual('react-redux');
  return {
    ...ActualReactRedux,
    useSelector: jest.fn().mockImplementation(() => {
      const amount = 1000,
        payback = 3;
      return { amount, payback };
    }),
    useDispatch: () => jest.fn(),
  };
});

test('Root renders correctly', () => {
  const tree = renderer.create(<Schedule />).toJSON();
  expect(tree).toMatchSnapshot();
});
