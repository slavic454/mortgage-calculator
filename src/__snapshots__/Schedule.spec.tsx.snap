// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="card-header"
      >
        <h1>
           
          Mortgage Calculator
           
        </h1>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="col-form-label-lg"
            htmlFor="amount"
          >
            ðŸ’µ Amount 
          </label>
          <input
            className="form-control-lg form-control"
            name="amount"
            onChange={[Function]}
            type="number"
            value="300000"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="col-form-label-lg"
            htmlFor="payback"
          >
            ðŸ“… Payment time in years
          </label>
          <input
            className="form-control-lg form-control"
            onChange={[Function]}
            type="number"
            value="20"
          />
        </div>
        <button
          aria-label={null}
          className="btn btn-primary btn-lg btn-block"
          disabled={false}
          onClick={[Function]}
        >
          Calculate
        </button>
      </form>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="card"
      >
        <div
          className="card-header"
        >
          <h1>
             
            Summary
             
          </h1>
        </div>
        <div
          className="card-group"
        >
          <div
            className="card-body"
          >
            <div>
               
              PieChart
               
            </div>
          </div>
          <div
            className="card-body"
          >
            <p
              className="card-text"
            >
              <h4>
                 Amount : 
                1000.00$
                 
              </h4>
            </p>
            <p
              className="card-text"
            >
              <h4>
                 Interest Rate: 
                3.50 %
                 
              </h4>
            </p>
            <p
              className="card-text"
            >
              <h4>
                 Total interests: 
                54.87$
                 
              </h4>
            </p>
            <p
              className="card-text"
            >
              <h4>
                Total sum : 
                1054.87$
                 
              </h4>
            </p>
          </div>
        </div>
      </div>
      <div
        className="card-header"
      >
        <h1>
           
          Amortization schedule
           
        </h1>
      </div>
      <div
        className="table-responsive"
      >
        <table
          className="table table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                #
              </th>
              <th>
                Payment
              </th>
              <th>
                Principal
              </th>
              <th>
                Interest
              </th>
              <th>
                Balance
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                scope="row"
              >
                1
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.39$
                 
              </td>
              <td>
                 
                2.92$
                 
              </td>
              <td>
                 
                973.61$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                2
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.46$
                 
              </td>
              <td>
                 
                2.84$
                 
              </td>
              <td>
                 
                947.15$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                3
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.54$
                 
              </td>
              <td>
                 
                2.76$
                 
              </td>
              <td>
                 
                920.61$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                4
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.62$
                 
              </td>
              <td>
                 
                2.69$
                 
              </td>
              <td>
                 
                894.00$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                5
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.69$
                 
              </td>
              <td>
                 
                2.61$
                 
              </td>
              <td>
                 
                867.30$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                6
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.77$
                 
              </td>
              <td>
                 
                2.53$
                 
              </td>
              <td>
                 
                840.53$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                7
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.85$
                 
              </td>
              <td>
                 
                2.45$
                 
              </td>
              <td>
                 
                813.68$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                8
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                26.93$
                 
              </td>
              <td>
                 
                2.37$
                 
              </td>
              <td>
                 
                786.75$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                9
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.01$
                 
              </td>
              <td>
                 
                2.29$
                 
              </td>
              <td>
                 
                759.74$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                10
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.09$
                 
              </td>
              <td>
                 
                2.22$
                 
              </td>
              <td>
                 
                732.66$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                11
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.17$
                 
              </td>
              <td>
                 
                2.14$
                 
              </td>
              <td>
                 
                705.49$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                12
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.24$
                 
              </td>
              <td>
                 
                2.06$
                 
              </td>
              <td>
                 
                678.25$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                13
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.32$
                 
              </td>
              <td>
                 
                1.98$
                 
              </td>
              <td>
                 
                650.92$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                14
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.40$
                 
              </td>
              <td>
                 
                1.90$
                 
              </td>
              <td>
                 
                623.52$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                15
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.48$
                 
              </td>
              <td>
                 
                1.82$
                 
              </td>
              <td>
                 
                596.04$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                16
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.56$
                 
              </td>
              <td>
                 
                1.74$
                 
              </td>
              <td>
                 
                568.47$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                17
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.64$
                 
              </td>
              <td>
                 
                1.66$
                 
              </td>
              <td>
                 
                540.83$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                18
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.72$
                 
              </td>
              <td>
                 
                1.58$
                 
              </td>
              <td>
                 
                513.10$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                19
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.81$
                 
              </td>
              <td>
                 
                1.50$
                 
              </td>
              <td>
                 
                485.30$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                20
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.89$
                 
              </td>
              <td>
                 
                1.42$
                 
              </td>
              <td>
                 
                457.41$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                21
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                27.97$
                 
              </td>
              <td>
                 
                1.33$
                 
              </td>
              <td>
                 
                429.44$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                22
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.05$
                 
              </td>
              <td>
                 
                1.25$
                 
              </td>
              <td>
                 
                401.39$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                23
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.13$
                 
              </td>
              <td>
                 
                1.17$
                 
              </td>
              <td>
                 
                373.26$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                24
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.21$
                 
              </td>
              <td>
                 
                1.09$
                 
              </td>
              <td>
                 
                345.05$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                25
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.30$
                 
              </td>
              <td>
                 
                1.01$
                 
              </td>
              <td>
                 
                316.75$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                26
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.38$
                 
              </td>
              <td>
                 
                0.92$
                 
              </td>
              <td>
                 
                288.37$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                27
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.46$
                 
              </td>
              <td>
                 
                0.84$
                 
              </td>
              <td>
                 
                259.91$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                28
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.54$
                 
              </td>
              <td>
                 
                0.76$
                 
              </td>
              <td>
                 
                231.37$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                29
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.63$
                 
              </td>
              <td>
                 
                0.67$
                 
              </td>
              <td>
                 
                202.74$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                30
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.71$
                 
              </td>
              <td>
                 
                0.59$
                 
              </td>
              <td>
                 
                174.03$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                31
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.79$
                 
              </td>
              <td>
                 
                0.51$
                 
              </td>
              <td>
                 
                145.24$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                32
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.88$
                 
              </td>
              <td>
                 
                0.42$
                 
              </td>
              <td>
                 
                116.36$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                33
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                28.96$
                 
              </td>
              <td>
                 
                0.34$
                 
              </td>
              <td>
                 
                87.40$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                34
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                29.05$
                 
              </td>
              <td>
                 
                0.25$
                 
              </td>
              <td>
                 
                58.35$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                35
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                29.13$
                 
              </td>
              <td>
                 
                0.17$
                 
              </td>
              <td>
                 
                29.22$
                 
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                36
              </th>
              <td>
                 
                29.30$
                 
              </td>
              <td>
                 
                29.22$
                 
              </td>
              <td>
                 
                0.09$
                 
              </td>
              <td>
                 
                -0.00$
                 
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
